File: app/src/main/java/bj/vinylbrowser/artist/ArtistController.java
package bj.vinylbrowser.artist;

import android.content.Context;

import bj.vinylbrowser.R;
import bj.vinylbrowser.epoxy.artist.MemberModel_;
import bj.vinylbrowser.epoxy.artist.UrlModel_;
import bj.vinylbrowser.epoxy.artist.ViewReleasesModel_;
import bj.vinylbrowser.epoxy.common.BaseController;
import bj.vinylbrowser.epoxy.common.DividerModel_;
import bj.vinylbrowser.epoxy.common.EmptySpaceModel_;
import bj.vinylbrowser.epoxy.common.LoadingModel_;
import bj.vinylbrowser.epoxy.common.RetryModel_;
import bj.vinylbrowser.epoxy.common.SubHeaderModel_;
import bj.vinylbrowser.model.artist.ArtistResult;
import bj.vinylbrowser.model.artist.ArtistWantedUrl;
import bj.vinylbrowser.model.artist.Member;
import bj.vinylbrowser.utils.ImageViewAnimator;
import bj.vinylbrowser.utils.analytics.AnalyticsTracker;

/**
 * Created by Josh Laird on 21/04/2017.
 */
public class ArtistController extends BaseController
{
    private final ArtistContract.View view;
    private final Context context;
    private ImageViewAnimator imageViewAnimator;
    private ArtistResult artistResult;
    private boolean loading = true;
    private boolean error;
    private AnalyticsTracker tracker;

    public ArtistController(ArtistContract.View view, Context context, ImageViewAnimator imageViewAnimator, AnalyticsTracker tracker)
    {
        this.view = view;
        this.context = context;
        this.imageViewAnimator = imageViewAnimator;
        this.tracker = tracker;
    }

    @Override
    protected void buildModels()
    {
        header
                .context(context)
                .imageViewAnimator(imageViewAnimator)
                .title(title)
                .subtitle(subtitle)
                .imageUrl(imageUrl)
                .addTo(this);

        new DividerModel_()
                .id("divider1")
                .addTo(this);

        new LoadingModel_()
                .imageViewAnimator(imageViewAnimator)
                .id("artist loading")
                .addIf(loading, this);

        new RetryModel_()
                .errorString("Unable to load Artist")
                .onClick(v -> view.retry())
                .id("error model")
                .addIf(error, this);

        if (artistResult != null)
        {
            if (artistResult.getMembers() != null && artistResult.getMembers().size() > 0)
            {
                new SubHeaderModel_()
                        .id("members subheader")
                        .subheader("Members")
                        .addTo(this);

                for (Member member : artistResult.getMembers())
                {
                    new MemberModel_()
                            .id("member" + artistResult.getMembers().indexOf(member))
                            .name(member.getName())
                            .onClick(v -> view.showMemberDetails(member.getName(), member.getId()))
                            .addTo(this);
                }

                new DividerModel_()
                        .id("divider2")
                        .addTo(this);
            }

            if (artistResult.getReleasesUrl() != null)
            {
                new ViewReleasesModel_()
                        .onViewReleasesClicked(v -> view.showArtistReleases(title, artistResult.getId()))
                        .id("viewReleases")
                        .addTo(this);

                new DividerModel_()
                        .id("releasesDivider")
                        .addTo(this);
            }

            if (artistResult.getArtistWantedUrls() != null && artistResult.getArtistWantedUrls().size() > 0)
            {
                new SubHeaderModel_()
                        .id("links subheader")
                        .subheader("Links")
                        .addTo(this);

                for (ArtistWantedUrl artistWantedUrl : artistResult.getArtistWantedUrls())
                {
                    new UrlModel_()
                            .id("wantedUrl" + artistResult.getArtistWantedUrls().indexOf(artistWantedUrl))
                            .onLinkClick(v -> view.openLink(artistWantedUrl.getUrl()))
                            .fontAwesomeCode(artistWantedUrl.getFontAwesomeString())
                            .friendlyText(artistWantedUrl.getDisplayText())
                            .hexCode(artistWantedUrl.getHexCode())
                            .addTo(this);
                }

            }
        }
        new EmptySpaceModel_()
                .id("empty space model")
                .addTo(this);
    }

    public void setArtist(ArtistResult artistResult)
    {
        this.artistResult = artistResult;
        if (artistResult.getImages() != null)
            imageUrl = artistResult.getImages().get(0).getResourceUrl();
        if (artistResult.getNameVariations() != null && artistResult.getNameVariations().size() > 0)
            title = artistResult.getNameVariations().get(0);
        subtitle = artistResult.getProfile();
        this.loading = false;
        this.error = false;
        requestModelBuild();
    }

    public void setLoading(boolean isLoading)
    {
        this.loading = isLoading;
        this.error = false;
        requestModelBuild();
    }

    public void setError(boolean isError)
    {
        this.error = isError;
        this.loading = false;
        tracker.send(context.getString(R.string.artist_activity), context.getString(R.string.artist_activity), context.getString(R.string.error), "fetching artist", "1");
        requestModelBuild();
    }
}

File: app/src/main/java/bj/vinylbrowser/artist/ArtistController.kt
package bj.vinylbrowser.artist

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import bj.vinylbrowser.R
import bj.vinylbrowser.artistreleases.ArtistReleasesController
import bj.vinylbrowser.common.BaseController
import bj.vinylbrowser.main.MainActivity
import bj.vinylbrowser.main.MainComponent
import bj.vinylbrowser.utils.analytics.AnalyticsTracker
import com.bluelinelabs.conductor.RouterTransaction
import com.bluelinelabs.conductor.changehandler.FadeChangeHandler
import kotlinx.android.synthetic.main.controller_recyclerview.view.*
import javax.inject.Inject

/**
 * Created by Josh Laird on 29/05/2017.
 */
class ArtistController(val title: String, val id: String) : BaseController(), ArtistContract.View {
    @Inject lateinit var tracker: AnalyticsTracker
    @Inject lateinit var presenter: ArtistPresenter
    @Inject lateinit var epxController: ArtistEpxController

    constructor(args: Bundle) : this(args.getString("title"), args.getString("id"))

    override fun setupComponent(mainComponent: MainComponent) {
        mainComponent
                .artistComponentBuilder()
                .artistModule(ArtistModule(this))
                .build()
                .inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup): View {
        val view = inflater.inflate(R.layout.controller_recyclerview, container, false)
        setupComponent((activity as MainActivity).mainComponent)
        setupToolbar(view.toolbar, "")
        setupRecyclerView(view.recyclerView, epxController)
        epxController.setTitle(title)
        epxController.requestModelBuild()
        presenter.fetchReleaseDetails(id)
        return view
    }

    override fun openLink(url: String?) {
        tracker.send(applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.clicked), url, "1")
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        activity?.startActivity(intent)
    }

    override fun onAttach(view: View) {
        super.onAttach(view)
        if (view.recyclerView.adapter == null)
            setupRecyclerView(view.recyclerView, epxController)
        tracker.send(applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.loaded), "onResume", "1")
    }

    override fun retry() {
        tracker.send(applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.clicked), "retry", "1")
        presenter.fetchReleaseDetails(id)
    }

    override fun showMemberDetails(name: String?, id: String?) {
        tracker.send(applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.clicked), "Show member details", "1")
        router.pushController(RouterTransaction.with(ArtistController(name!!, id!!))
                .popChangeHandler(FadeChangeHandler())
                .pushChangeHandler(FadeChangeHandler())
                .tag("ArtistController"))
    }

    override fun showArtistReleases(title: String, id: String) {
        tracker.send(applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.artist_activity), applicationContext?.getString(R.string.clicked), "show artist releases", "1")
        router.pushController(RouterTransaction.with(ArtistReleasesController(title, id))
                .popChangeHandler(FadeChangeHandler())
                .pushChangeHandler(FadeChangeHandler())
                .tag("ArtistReleasesController"))
    }

    override fun onRestoreViewState(view: View, savedViewState: Bundle) {
        super.onRestoreViewState(view, savedViewState)
        epxController.setArtist(savedViewState.getParcelable("artist"))
    }

    override fun onSaveViewState(view: View, outState: Bundle) {
        outState.putParcelable("artist", epxController.artistResult)
        super.onSaveViewState(view, outState)
    }
}